Weather,clouds and panorama layer for google maps

Gitek1 opened the pull request:
Added checkbox for weather, clouds and panorama layer.

Gitek1 committed:
Panel added weather, clouds and panorama checkbox.

Gitek1 committed:
Two stupid mistake fixed.

Gitek1 commented:
Check it.

mitar commented:
I don't see clouds working.

Why all labels on the map are hidden when you enable weather? I would leave labels on, even if things become more crowded, because layers in general it would be useful to have labels on.

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -29,6 +29,18 @@ $(document).ready(function () {
     var map = new google.maps.Map($('#map').get(0), options);
 
+    // Necessary libraries are included with googleapis in piplmesh/panels/map/templates/panel/map/panel.html
+    var map_layers = new Object;
```
All this is not good way to define things. Order should be predictable. So use list. Then define each layer as one entry there. Something like:

```
var map_layers = [
    {
        'label': gettext("Weather"),
        'layer': ew google.maps.weather.WeatherLayer(...)
    },
    {
        ...
    }
];
```

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -29,6 +29,18 @@ $(document).ready(function () {
 
+    // Necessary libraries are included with googleapis in piplmesh/panels/map/templates/panel/map/panel.html
+    var map_layers = new Object;
+    var map_layers_labels = new Object;
```
In general, new empty object can be done with `{}`. Much nicer code.

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -29,6 +29,18 @@ $(document).ready(function () {
+    var map_layers_labels = new Object;
+    map_layers_labels.weather = gettext('Weather');
+    map_layers.weather = new google.maps.weather.WeatherLayer({
+        temperatureUnits: google.maps.weather.TemperatureUnit.CELSIUS
```
Use `'` around keys.

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -29,6 +29,18 @@ $(document).ready(function () {
+    map_layers.weather = new google.maps.weather.WeatherLayer({
+        temperatureUnits: google.maps.weather.TemperatureUnit.CELSIUS
+    });
+    map_layers_labels.clouds = gettext('Clouds');
```
User-exposed strings should be in `"`.

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -93,6 +129,12 @@ $(document).ready(function () {
         $('#advanced-map-container').show();
+        $('<div/>').prop('id', 'map-layers').appendTo('#advanced-map');
+        // Map layer options should be added after #map-layers is created
+        for (var layer in map_layers) {
```
Use jQuery `.each`.

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -60,6 +72,30 @@ $(document).ready(function () {
+                'name': 'map-layer-' + layer
+            }).appendTo('#map-layers');
+            $('<label/>').prop('for', 'map-layer-' + layer).text(label).appendTo('#map-layers');
+            $('<br/>').appendTo('#map-layers');
```
Why is `<br/>` needed? In general this is something which is related to design. So you can do that with CSS. Probably you just want one div around and it will be same effect?

Gitek1 committed:
changed how map_layers are defined

Gitek1 commented:
Fixed.

I see clouds when zoomed out. I removed them in last commit, I guess they are not that usable.

I think labels cant be shown in weather mode, there is no option in documentation.
Documentation says ( https://developers.google.com/maps/documentation/javascript/layers#WeatherLayer ):
"When the weather layer is enabled and the zoom level is set to 12 or lower, administrative labels, such as the names of streets or regions, will be disabled on your map"

mitar commented:
I would have cloud layer. If not, it looks interesting. :-) It is already optional. But also later on we can make people be able to configure which layers to see.

mitar commented:
OK, then if this is how weather layer behaves, let it be.

Can you maybe add a tooltip over the layer entry in the list on the right? Then we could have a tooltip like "Hides labels on zoom level 12 or lower."

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -29,6 +29,22 @@ $(document).ready(function () {
+            'id': 'weather',
+            'label': gettext("Weather"),
+            'layer': new google.maps.weather.WeatherLayer({
+                'temperatureUnits': google.maps.weather.TemperatureUnit.CELSIUS
```
Is there a way to get this from user's locale? If not, add TODO that this should be user-configurable.

mitar replied:
(You always have to think about such assumptions. World is global today! Products/services should also be!)

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -60,6 +76,30 @@ $(document).ready(function () {
+    }
+
+    function setLayerVisibility(id, layer) {
+        if ($('#map-layer-' + id).attr('checked')) {
```
Use `prop`, not `attr`. [Read here](http://api.jquery.com/prop/).

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -93,6 +133,12 @@ $(document).ready(function () {
         $('#advanced-map-container').show();
+        $('<div/>').prop('id', 'map-layers').appendTo('#advanced-map');
+        // Map layer options should be added after #map-layers is created
+        $.each(map_layers, function(key, option) {
```
Rename `option` to `layer`. You iterate over layers, so each element is a layer, no? Or rename `map_layers` to `map_layers_options`.

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -60,6 +76,30 @@ $(document).ready(function () {
     }
 
+    function addMapLayerOption(id, label, layer) {
+        if($('#map-layers').lenght != 0) {
```
Space before `(`.

Have you run this code? Because it is not `lenght` but `length`.

In general it is better:

```
if ($('#map-layers').length) {
```

Or even:

```
if ($('#map-layers')) {
```

mitar replied:
The question is, is it really necessary? In general it is better sometimes to get an error if something is not as expected (like, there is no #map-layers).

But in this case you can also leave to jQuery to simply not do anything, when you want to append to it. And this is it. No need for special check.

Gitek1 replied:
Yes, I ran the code and it worked, not even a error in chrome console (?), that's why I didn't notice the typo, so now I just removed it

mitar replied:
There was no error because the value was always undefined, which is always false. So my question is, does the logic of your program worked. If it works with this being always false then the question which appears is: do we really need this whole block of code inside the if. :-)

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -60,6 +76,30 @@ $(document).ready(function () {
 
+    function addMapLayerOption(id, label, layer) {
+        if($('#map-layers').lenght != 0) {
+            $('<div/>').prop('id','map-option-' + id).appendTo('#map-layers');
```
Comma after `,`.

mitar replied:
Do you really need ID on this div? You could just store jQuery div object into a variable and use that as a target of `appendTo`. No need to do the query against ID there.

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -93,6 +133,12 @@ $(document).ready(function () {
+        $('<div/>').prop('id', 'map-layers').appendTo('#advanced-map');
+        // Map layer options should be added after #map-layers is created
+        $.each(map_layers, function(key, option) {
+            addMapLayerOption(option.id, option.label, option.layer);
```
Why not simply pass `option` object?

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -60,6 +76,30 @@ $(document).ready(function () {
+                'id': 'map-layer-' + id,
+                'type': 'checkbox',
+                'name': 'map-layer-' + id
+            }).appendTo('#map-option-' + id);
```
So you could just append to object stored in a variable here.

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -60,6 +76,30 @@ $(document).ready(function () {
+                'type': 'checkbox',
+                'name': 'map-layer-' + id
+            }).appendTo('#map-option-' + id);
+            $('<label/>').prop('for', 'map-layer-' + id).text(label).appendTo('#map-option-' + id);
```
So you could just append to object stored in a variable here, at the end.

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -60,6 +76,30 @@ $(document).ready(function () {
+                'name': 'map-layer-' + id
+            }).appendTo('#map-option-' + id);
+            $('<label/>').prop('for', 'map-layer-' + id).text(label).appendTo('#map-option-' + id);
+            $('#map-layer-' + id).click(function (event) {
```
Hm, is click the best one? Or should be maybe instead change? Or both? Maybe both?

mitar replied:
And same here. You should chain jQuery calls together. You create input above, just attach also event handler there.

Gitek1 replied:
I think both would cancel each other out. I used change in the last commit.

mitar replied:
So I remember from past that there are some issues with checkboxes. So because you can change them also through keyboard (navigate to it with tab and press space). So you should make it work also with that.

Anyway, make it work with mouse and keyboard.

Hm, or it was maybe input box and keypress and change. :-)

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -93,6 +133,12 @@ $(document).ready(function () {
         $('#advanced-map-container').hide();
         $('#map').detach().prependTo('#advanced-map-container');
         $('#advanced-map-container').show();
+        $('<div/>').prop('id', 'map-layers').appendTo('#advanced-map');
```
Same here, do you really need to set ID, or you could store div object into variable and pass it to `addMapLayerOption`?

Gitek1 replied:
I kept this one (#map-layers) because I used it in css to style the div.

mitar replied:
You can keep it defined, but don't use it in the code, if you can store it into a variable.

mitar replied:
(Of course is variable is easily passed around. You do not have to complicate too much here. I just wanted to point out that this can be done instead of using IDs everywhere automatically.)

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -60,6 +76,30 @@ $(document).ready(function () {
+        }
+    }
+
+    function setLayerVisibility(id, layer) {
```
Similar here, you could pass the object?

So, the problem with all this relaying on IDs is that sometimes you could set invalid IDs, or that there are duplicates. Much easier is to do the query once, or construct the object, and then pass it around. Also from performance stand point (but here we do not really have that issue).

Gitek1 committed:
renamed option to map_layer, used variable insteas div id for appendTo and if/else

Gitek1 commented:
Fixed.

I don't think weather option needs a tooltip, clouds may be more problematic.

mitar commented:
> I don't think weather option needs a tooltip, clouds may be more problematic.

A tooltip explaining why labels disappeared? I think this is useful because if people enable more layers at the same time, to understand which one hide labels.

What is the issue with clouds?

mitar commented:
One more thing I found out. It is better to use `.attr` when defining new DOM in the code, and `.prop` when you want to set or access current values after DOM was in use.

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -29,6 +29,28 @@ $(document).ready(function () {
 
     var map = new google.maps.Map($('#map').get(0), options);
 
+    // TODO: This should be user-configurable
```
??? Why TODO here? It should be down there with celsius line.

Gitek1 replied:
I understood that "map_layers" should be configurable, I fixed this in last commit.

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -60,6 +82,27 @@ $(document).ready(function () {
 
+    function addMapLayerOption(map_layer) {
+        var checkbox_container = $('<div/>').appendTo('#map-layers');
+        var checkbox = $('<input/>').prop({
```
So `attr` here.

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -60,6 +82,27 @@ $(document).ready(function () {
+        }).change(function (event) {
+            setLayerVisibility(checkbox, map_layer);
+        }).appendTo(checkbox_container);
+        $('<label/>').prop('for', 'map-layer-' + map_layer.id).text(map_layer.label).appendTo(checkbox_container);
```
And here.

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -60,6 +82,27 @@ $(document).ready(function () {
+    }
+
+    function setLayerVisibility(checkbox, map_layer) {
+        if ((checkbox).prop('checked')) {
```
But here `prop` is good.

Why `()` around `checkbox`?

mitar commented on piplmesh/panels/map/static/piplmesh/panel/map/panel.js:
```diff
@@ -93,6 +136,12 @@ $(document).ready(function () {
         $('#advanced-map-container').show();
+        $('<div/>').prop('id', 'map-layers').appendTo('#advanced-map');
+        // Map layer options should be added after #map-layers is created
+        $.each(map_layers, function(key, map_layer) {
```
`key`? What is `key`? Rename it to mean something.

Gitek1 replied:
This function takes index key and value from map_layers. I renamed key to index in last commit.

Gitek1 committed:
replaced .prop with .attr, fixed small mistakes

Gitek1 commented:
Fixed.

The thing with clouds is that they don't show, even if enabled, when you are zoomed in. When user clicks the weather something happens no matter what zoom level the user is on.

mitar commented:
So it is not possible to configure tooltip ("title" DOM attribute) for clouds and weather. Something like:
- _Visible on zoom level X and above._
- _Hides map labels on zoom level X and above._

mitar commented:
Otherwise good.

Gitek1 committed:
added title attribute

Gitek1 commented:
Added tooltips
